cli/src/commands/parallelize.rs

```diff
@@ -83,61 +82,71 @@ pub(crate) fn cmd_parallelize(
+                new_parents.extend(grand_parents.clone());
+            } else {
+                new_parents.push(old_parent.clone());
+            }
         }
+        new_target_parents.insert(commit.id().clone(), new_parents);
     }
 
-    // Rebase the target commits onto the parents of the root commit.
-    // We already checked that the roots have the same parents, so we can just
-    // use the first one.
-    let target_commits = target_commits
-        .into_iter()
-        // We need to reverse the iterator so that when we rebase the target
-        // commits they will appear in the same relative order in `jj log` that
-        // they were in before being parallelized. After reversing, the commits
-        // are ordered with children before parents.
-        .rev()
-        .collect_vec();
-    rebase_descendants(
-        &mut tx,
+    // If a commit outside the target set has a commit in the target set as parent,
+    // then - after the transformation - it should also have that commit's
+    // parents as direct parents, if those commits are also in the target set.
+    let mut new_child_parents: HashMap<CommitId, IndexSet<CommitId>> = HashMap::new();
+    for commit in target_commits.iter().rev() {
+        let mut new_parents = IndexSet::new();
+        for old_parent in commit.parent_ids() {
+            if let Some(parents) = new_child_parents.get(old_parent) {
+                new_parents.extend(parents.iter().cloned());
+            }
+        }
+        new_parents.insert(commit.id().clone());
+        new_child_parents.insert(commit.id().clone(), new_parents);
+    }
+
+    tx.mut_repo().transform_descendants(
         command.settings(),
-        new_parents,
-        &target_commits,
-        Default::default(),
+        target_commits.iter().ids().cloned().collect_vec(),
+        |mut rewriter| {
+            // Commits in the target set do not depend on each other but they still depend
+            // on other parents
+            if let Some(new_parents) = new_target_parents.get(rewriter.old_commit().id()).cloned() {
+                let new_parents = rewriter.mut_repo().new_parents(&new_parents);
+                rewriter.set_new_parents(new_parents);
+            } else if rewriter
+                .old_commit()
+                .parent_ids()
+                .iter()
+                .any(|id| new_child_parents.contains_key(id))
+            {
+                let mut new_parents = vec![];
+                for parent in rewriter.old_commit().parent_ids() {
+                    if let Some(parents) = new_child_parents.get(parent) {
+                        new_parents.extend(parents.clone());
+                    } else {
+                        new_parents.push(parent.clone());
+                    }
+                }
+                let new_parents = rewriter.mut_repo().new_parents(&new_parents);
+                rewriter.set_new_parents(new_parents);
```

Comment:
nit: might be better to add rewriter method that looks up "rewritten" new parents and set them, but I'm not sure.

Reply:
Good idea. Done.